specs/WebAuthenticationRequested.md

```diff
@@ -0,0 +1,276 @@
+-->
+
+
+# Background
+By default HTTP basic authentication and NTLM authentication requests inside WebView2 show the authentication UI, which is a dialog prompt in which the user can type in user name and password credentials just like in the Edge browser. We have been requested by WebView2 app developers to provide finer granularity for managing HTTP Basic/NTLM authentications inside WebView2, including the ability to hide the login UI and provide credentials.
+
+
+# Description
+We propose a new event for WebView2, CoreWebView2BasicAuthenticationRequested that will allow developers to listen on and override the HTTP Basic authentication requests in WebView2. When there is a HTTP Basic/NTLM authentication request in WebView2, the developer will have a choice to:
+  1) Provide credentials
+  2) Cancel the login altogether
+  3) Ask the user for credentials via the default login prompt
+We also propose CoreWebView2BasicAuthenticationResponse, the runtime class that represents the app's response with credentials to the basic authentication request.
+
+# Examples
+## Provide credentials
+The developer can provide the authentication credentials on behalf of the user when it encounters the Basic authentication request. In this case, the default login dialog prompt will no longer be shown to the user. If the developer provided credentials are wrong, the server may keep responding with Unauthorized, which will lead to an infinite loop so the developer should pay attention to this.
+
+```cpp
+
+    CHECK_HRESULT(webview2->add_BasicAuthenticationRequested(
+        Callback<ICoreWebView2BasicAuthenticationRequestedEventHandler>(
+            [this](
+                ICoreWebView2* sender,
+                ICoreWebView2BasicAuthenticationRequestedEventArgs* args)
+            {
+                wil::com_ptr<ICoreWebView2Deferral> deferral;
+
+                // Ensure the event args are retained for the lambda.
+                wil::com_ptr<ICoreWebView2BasicAuthenticationRequestedEventArgs> web_auth_args = args;
+                args->GetDeferral(&deferral);
+                ShowCustomLoginUI().then([web_auth_args, deferral](LPCWSTR userName, LPCWSTR password)
+                {
+                    wil::com_ptr<ICoreWebView2BasicAuthenticationResponse> basicAuthenticationResponse;
+                    CHECK_HRESULT(web_auth_args->get_Response(&basicAuthenticationResponse));
+                    CHECK_HRESULT(basicAuthenticationResponse->put_UserName(userName));
+                    CHECK_HRESULT(basicAuthenticationResponse->put_Password(password));
+                    deferral->Complete();
+                });
+                
+                return S_OK;
+            })
+            .Get(),
+        &m_BasicAuthenticationRequestedToken));
+```
+
+```c#
+webView.CoreWebView2.BasicAuthenticationRequested += delegate (object sender, CoreWebView2BasicAuthenticationRequestedEventArgs args)
+{
+    CoreWebView2Deferral deferral = args.GetDeferral();
```

Comment:
Does CoreWebView2Deferral support IClosable/IDisposable? If so, then you can write

```csharp
using (CoreWebView2Deferral deferral = args.GetDeferral())
{
    Credential credential = await ShowCustomLoginUIAsync();
    args.Response.UserName = credential.UserName;
    args.Response.Password = credential.Password;
    // auto-disposed at end of scope, even if an exception occurs
}
```

Reply:
(If it doesn't, it should)

Reply:
It does (should) please update